### DISCUSSION TITLE
Automatically create database when tenant setup

### DISCUSSION THREAD

**Discussion by stantoxt** [OrchardCMS/OrchardCore#15850] [Q&A]
Why OrchardCore can not automatically create database when tenant is setup?

**Comment by sebastienros** [comment]
Because it would require a connection string that point to a different database than the one the site will use, and an account that also has these permissions. However for Sqlite it's not required, so you can already do this.

**Comment by stantoxt** [comment]
Thanks @sebastienros . I think oc that use relational sql server should use connetion string for default tenant(oc first time setup), but connection config should be unaware for undefault tenant (real tenant who wants to use oc cms) for relational sql server like for sqlite.Many other website builders dont need connection string for tenant.

**Comment by hishamco** [comment]
@stantoxt is that mean you all the tenants data to be stored in one database?!! what if I want each tenant has it's own database, each one may has different database provider

**Comment by ns8482e** [comment]
Even in case of using same database for multiple tenant - at some point you may need  to have separate database - like first 50 tenants are in Database A and 50 to 100 are in database B.

**Comment by stantoxt** [comment]
The default tenant has ability to create database,or enable a database-creation module to let the default tenant do this either in one database or separate database. A real tenant can setup its own web without database configuration (like for sqlite) from client-end (not-admin-end) .

**Comment by hishamco** [comment]
> A real tenant can setup its own web without database configuration (like for sqlite) from client-end (not-admin-end) .

Could you please elaborate

**Comment by stantoxt** [comment]
@hishamco I am so sorry to reply this now.
> > A real tenant can setup its own web without database configuration (like for sqlite) from client-end (not-admin-end) .
> 
> Could you please elaborate

For now,when we use sqlserver to set up OC, it will failed if there is no existing db. Maybe oc could automatically create db for a tenant setup. In a website builder platform, the tenants don't need to setup the site with db connection config, this db config is ok for developers,but it is hard to operate for customers without dev experience.

**Comment by hishamco** [comment]
> Maybe oc could automatically create db for a tenant setup

Some CMSs doing that, but I think connection to existing database may better, coz the user may don't have permission for creating a database

**Comment by stantoxt** [comment]
Sqlite is good enough for db creation automatically. And security is important.But I think it should be a feature.

**Comment by infofromca** [comment]
> The default tenant has ability to create database,or enable a database-creation module to let the default tenant do this either in one database or separate database. A real tenant can setup its own web without database configuration (like for sqlite) from client-end (not-admin-end) .

I think we can set  a  default connection string which is from default tenant, then if you need diff. db, set a custom indication(settings)

**Comment by stantoxt** [comment]
It's great.For now we can't create database from tenant setup.

**Comment by hishamco** [comment]
> Sqlite is good enough for db creation automatically. And security is important.But I think it should be a feature.

What the feature that you refer to?

> It's great.For now we can't create database from tenant setup.

Why?!!

**Comment by stantoxt** [comment]
> > Sqlite is good enough for db creation automatically. And security is important.But I think it should be a feature.
> 
> What the feature that you refer to?

It's about this.
> Some CMSs doing that, but I think connection to existing database may better, coz the user may don't have permission for creating a database

The saas host institution should supply database storage, and the database storage can't be limited to sqlite-like database. Relational Database should be an option,if a tenant use this feature they can create their own database and setup web with the database.

> > It's great.For now we can't create database from tenant setup.
> 
> Why?!!

If oc use relational sql server like postgresql or sql server, tenant must fill in connectionstring field with existing database in setup action. but a user don't need to create an empty database and then migrate the database from the code when using efcore code first.
